File: test2/ng/projects/test2/src/lib/front-repo.service.ts
import { Injectable } from '@angular/core'
import { HttpClient, HttpHeaders } from '@angular/common/http'

import { Observable, combineLatest, BehaviorSubject, of } from 'rxjs'

// insertion point sub template for services imports
import { ADB } from './a-db'
import { AService } from './a.service'

import { BDB } from './b-db'
import { BService } from './b.service'


// FrontRepo stores all instances in a front repository (design pattern repository)
export class FrontRepo { // insertion point sub template
  As_array = new Array<ADB>() // array of repo instances
  As = new Map<number, ADB>() // map of repo instances
  As_batch = new Map<number, ADB>() // same but only in last GET (for finding repo instances to delete)

  Bs_array = new Array<BDB>() // array of repo instances
  Bs = new Map<number, BDB>() // map of repo instances
  Bs_batch = new Map<number, BDB>() // same but only in last GET (for finding repo instances to delete)


  // getArray allows for a get function that is robust to refactoring of the named struct name
  // for instance frontRepo.getArray<Astruct>( Astruct.GONGSTRUCT_NAME), is robust to a refactoring of Astruct identifier
  // contrary to frontRepo.Astructs_array which is not refactored when Astruct identifier is modified
  getArray<Type>(gongStructName: string): Array<Type> {
    switch (gongStructName) {
      // insertion point
      case 'A':
        return this.As_array as unknown as Array<Type>
      case 'B':
        return this.Bs_array as unknown as Array<Type>
      default:
        throw new Error("Type not recognized");
    }
  }

  // getMap allows for a get function that is robust to refactoring of the named struct name
  getMap<Type>(gongStructName: string): Map<number, Type> {
    switch (gongStructName) {
      // insertion point
      case 'A':
        return this.As_array as unknown as Map<number, Type>
      case 'B':
        return this.Bs_array as unknown as Map<number, Type>
      default:
        throw new Error("Type not recognized");
    }
  }
}

// the table component is called in different ways
//
// DISPLAY or ASSOCIATION MODE
//
// in ASSOCIATION MODE, it is invoked within a diaglo and a Dialog Data item is used to
// configure the component
// DialogData define the interface for information that is forwarded from the calling instance to 
// the select table
export class DialogData {
  ID: number = 0 // ID of the calling instance

  // the reverse pointer is the name of the generated field on the destination
  // struct of the ONE-MANY association
  ReversePointer: string = "" // field of {{Structname}} that serve as reverse pointer
  OrderingMode: boolean = false // if true, this is for ordering items

  // there are different selection mode : ONE_MANY or MANY_MANY
  SelectionMode: SelectionMode = SelectionMode.ONE_MANY_ASSOCIATION_MODE

  // used if SelectionMode is MANY_MANY_ASSOCIATION_MODE
  //
  // In Gong, a MANY-MANY association is implemented as a ONE-ZERO/ONE followed by a ONE_MANY association
  // 
  // in the MANY_MANY_ASSOCIATION_MODE case, we need also the Struct and the FieldName that are
  // at the end of the ONE-MANY association
  SourceStruct: string = ""  // The "Aclass"
  SourceField: string = "" // the "AnarrayofbUse"
  IntermediateStruct: string = "" // the "AclassBclassUse" 
  IntermediateStructField: string = "" // the "Bclass" as field
  NextAssociationStruct: string = "" // the "Bclass"

  GONG__StackPath: string = ""
}

export enum SelectionMode {
  ONE_MANY_ASSOCIATION_MODE = "ONE_MANY_ASSOCIATION_MODE",
  MANY_MANY_ASSOCIATION_MODE = "MANY_MANY_ASSOCIATION_MODE",
}

//
// observable that fetch all elements of the stack and store them in the FrontRepo
//
@Injectable({
  providedIn: 'root'
})
export class FrontRepoService {

  GONG__StackPath: string = ""

  httpOptions = {
    headers: new HttpHeaders({ 'Content-Type': 'application/json' })
  };

  //
  // Store of all instances of the stack
  //
  frontRepo = new (FrontRepo)

  constructor(
    private http: HttpClient, // insertion point sub template 
    private aService: AService,
    private bService: BService,
  ) { }

  // postService provides a post function for each struct name
  postService(structName: string, instanceToBePosted: any) {
    let service = this[structName.toLowerCase() + "Service" + "Service" as keyof FrontRepoService]
    let servicePostFunction = service[("post" + structName) as keyof typeof service] as (instance: typeof instanceToBePosted) => Observable<typeof instanceToBePosted>

    servicePostFunction(instanceToBePosted).subscribe(
      instance => {
        let behaviorSubject = instanceToBePosted[(structName + "ServiceChanged") as keyof typeof instanceToBePosted] as unknown as BehaviorSubject<string>
        behaviorSubject.next("post")
      }
    );
  }

  // deleteService provides a delete function for each struct name
  deleteService(structName: string, instanceToBeDeleted: any) {
    let service = this[structName.toLowerCase() + "Service" as keyof FrontRepoService]
    let serviceDeleteFunction = service["delete" + structName as keyof typeof service] as (instance: typeof instanceToBeDeleted) => Observable<typeof instanceToBeDeleted>

    serviceDeleteFunction(instanceToBeDeleted).subscribe(
      instance => {
        let behaviorSubject = instanceToBeDeleted[(structName + "ServiceChanged") as keyof typeof instanceToBeDeleted] as unknown as BehaviorSubject<string>
        behaviorSubject.next("delete")
      }
    );
  }

  // typing of observable can be messy in typescript. Therefore, one force the type
  observableFrontRepo: [
    Observable<null>, // see below for the of(null) observable
    // insertion point sub template 
    Observable<ADB[]>,
    Observable<BDB[]>,
  ] = [
      // Using "combineLatest" with a placeholder observable.
      //
      // This allows the typescript compiler to pass when no GongStruct is present in the front API
      //
      // The "of(null)" is a "meaningless" observable that emits a single value (null) and completes.
      // This is used as a workaround to satisfy TypeScript requirements and the "combineLatest" 
      // expectation for a non-empty array of observables.
      of(null), // 
      // insertion point sub template
      this.aService.getAs(this.GONG__StackPath),
      this.bService.getBs(this.GONG__StackPath),
    ];

  //
  // pull performs a GET on all struct of the stack and redeem association pointers 
  //
  // This is an observable. Therefore, the control flow forks with
  // - pull() return immediatly the observable
  // - the observable observer, if it subscribe, is called when all GET calls are performs
  pull(GONG__StackPath: string = ""): Observable<FrontRepo> {

    this.GONG__StackPath = GONG__StackPath

    this.observableFrontRepo = [
      of(null), // see above for justification
      // insertion point sub template
      this.aService.getAs(this.GONG__StackPath),
      this.bService.getBs(this.GONG__StackPath),
    ]

    return new Observable<FrontRepo>(
      (observer) => {
        combineLatest(
          this.observableFrontRepo
        ).subscribe(
          ([
            ___of_null, // see above for the explanation about of
            // insertion point sub template for declarations 
            as_,
            bs_,
          ]) => {
            // Typing can be messy with many items. Therefore, type casting is necessary here
            // insertion point sub template for type casting 
            var as: ADB[]
            as = as_ as ADB[]
            var bs: BDB[]
            bs = bs_ as BDB[]

            // 
            // First Step: init map of instances
            // insertion point sub template for init 
            // init the array
            this.frontRepo.As_array = as

            // clear the map that counts A in the GET
            this.frontRepo.As_batch.clear()

            as.forEach(
              a => {
                this.frontRepo.As.set(a.ID, a)
                this.frontRepo.As_batch.set(a.ID, a)
              }
            )

            // clear as that are absent from the batch
            this.frontRepo.As.forEach(
              a => {
                if (this.frontRepo.As_batch.get(a.ID) == undefined) {
                  this.frontRepo.As.delete(a.ID)
                }
              }
            )

            // sort As_array array
            this.frontRepo.As_array.sort((t1, t2) => {
              if (t1.Name > t2.Name) {
                return 1;
              }
              if (t1.Name < t2.Name) {
                return -1;
              }
              return 0;
            });

            // init the array
            this.frontRepo.Bs_array = bs

            // clear the map that counts B in the GET
            this.frontRepo.Bs_batch.clear()

            bs.forEach(
              b => {
                this.frontRepo.Bs.set(b.ID, b)
                this.frontRepo.Bs_batch.set(b.ID, b)
              }
            )

            // clear bs that are absent from the batch
            this.frontRepo.Bs.forEach(
              b => {
                if (this.frontRepo.Bs_batch.get(b.ID) == undefined) {
                  this.frontRepo.Bs.delete(b.ID)
                }
              }
            )

            // sort Bs_array array
            this.frontRepo.Bs_array.sort((t1, t2) => {
              if (t1.Name > t2.Name) {
                return 1;
              }
              if (t1.Name < t2.Name) {
                return -1;
              }
              return 0;
            });


            // 
            // Second Step: redeem pointers between instances (thanks to maps in the First Step)
            // insertion point sub template for redeem 
            as.forEach(
              a => {
                // insertion point sub sub template for ONE-/ZERO-ONE associations pointers redeeming
                // insertion point for pointer field B redeeming
                {
                  let _b = this.frontRepo.Bs.get(a.APointersEncoding.BID.Int64)
                  if (_b) {
                    a.B = _b
                  }
                }

                // insertion point for redeeming ONE-MANY associations
              }
            )
            bs.forEach(
              b => {
                // insertion point sub sub template for ONE-/ZERO-ONE associations pointers redeeming

                // insertion point for redeeming ONE-MANY associations
                // insertion point for slice of pointer field A.Bs redeeming
                // to be removed
                {
                  let _id = b.BPointersEncoding.A_BsDBID.Int64
                  let _a = this.frontRepo.As.get(_id)
                  if (_a) {
                    if (_a.Bs == undefined) {
                      _a.Bs = new Array<BDB>()
                    }
                    _a.Bs.push(b)
                    if (b.BPointersEncoding.A_Bs_reverse == undefined) {
                      b.BPointersEncoding.A_Bs_reverse = _a
                    }
                  }
                }
              }
            )

            // 
            // Third Step: sort arrays (slices in go) according to their index
            // insertion point sub template for redeem 
            as.forEach(
              a => {
                // insertion point for sorting
                // to be removed
                a.Bs?.sort((t1, t2) => {
                  if (t1.BPointersEncoding.A_BsDBID_Index.Int64 > t2.BPointersEncoding.A_BsDBID_Index.Int64) {
                    return 1;
                  }
                  if (t1.BPointersEncoding.A_BsDBID_Index.Int64 < t2.BPointersEncoding.A_BsDBID_Index.Int64) {
                    return -1;
                  }
                  return 0;
                })

              }
            )
            bs.forEach(
              b => {
                // insertion point for sorting
              }
            )

            // hand over control flow to observer
            observer.next(this.frontRepo)
          }
        )
      }
    )
  }

  // insertion point for pull per struct 

  // APull performs a GET on A of the stack and redeem association pointers 
  APull(): Observable<FrontRepo> {
    return new Observable<FrontRepo>(
      (observer) => {
        combineLatest([
          this.aService.getAs(this.GONG__StackPath)
        ]).subscribe(
          ([ // insertion point sub template 
            as,
          ]) => {
            // init the array
            this.frontRepo.As_array = as

            // clear the map that counts A in the GET
            this.frontRepo.As_batch.clear()

            // 
            // First Step: init map of instances
            // insertion point sub template 
            as.forEach(
              a => {
                this.frontRepo.As.set(a.ID, a)
                this.frontRepo.As_batch.set(a.ID, a)

                // insertion point for redeeming ONE/ZERO-ONE associations
                // insertion point for pointer field B redeeming
                {
                  let _b = this.frontRepo.Bs.get(a.APointersEncoding.BID.Int64)
                  if (_b) {
                    a.B = _b
                  }
                }

                // insertion point for redeeming ONE-MANY associations
              }
            )

            // clear as that are absent from the GET
            this.frontRepo.As.forEach(
              a => {
                if (this.frontRepo.As_batch.get(a.ID) == undefined) {
                  this.frontRepo.As.delete(a.ID)
                }
              }
            )

            // 
            // Second Step: redeem pointers between instances (thanks to maps in the First Step)
            // insertion point sub template 

            // hand over control flow to observer
            observer.next(this.frontRepo)
          }
        )
      }
    )
  }

  // BPull performs a GET on B of the stack and redeem association pointers 
  BPull(): Observable<FrontRepo> {
    return new Observable<FrontRepo>(
      (observer) => {
        combineLatest([
          this.bService.getBs(this.GONG__StackPath)
        ]).subscribe(
          ([ // insertion point sub template 
            bs,
          ]) => {
            // init the array
            this.frontRepo.Bs_array = bs

            // clear the map that counts B in the GET
            this.frontRepo.Bs_batch.clear()

            // 
            // First Step: init map of instances
            // insertion point sub template 
            bs.forEach(
              b => {
                this.frontRepo.Bs.set(b.ID, b)
                this.frontRepo.Bs_batch.set(b.ID, b)

                // insertion point for redeeming ONE/ZERO-ONE associations

                // insertion point for redeeming ONE-MANY associations
                // insertion point for slice of pointer field A.Bs redeeming
                // to be removed
                {
                  let _id = b.BPointersEncoding.A_BsDBID.Int64
                  let _a = this.frontRepo.As.get(_id)
                  if (_a) {
                    if (_a.Bs == undefined) {
                      _a.Bs = new Array<BDB>()
                    }
                    _a.Bs.push(b)
                    if (b.BPointersEncoding.A_Bs_reverse == undefined) {
                      b.BPointersEncoding.A_Bs_reverse = _a
                    }
                  }
                }
              }
            )

            // clear bs that are absent from the GET
            this.frontRepo.Bs.forEach(
              b => {
                if (this.frontRepo.Bs_batch.get(b.ID) == undefined) {
                  this.frontRepo.Bs.delete(b.ID)
                }
              }
            )

            // 
            // Second Step: redeem pointers between instances (thanks to maps in the First Step)
            // insertion point sub template 

            // hand over control flow to observer
            observer.next(this.frontRepo)
          }
        )
      }
    )
  }
}

// insertion point for get unique ID per struct 
export function getAUniqueID(id: number): number {
  return 31 * id
}
export function getBUniqueID(id: number): number {
  return 37 * id
}

File: go/angular/front-repo.service.go
package angular

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"github.com/fullstack-lang/gong/go/models"
)

const NgLibFrontRepoServiceTemplate = `import { Injectable } from '@angular/core'
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http'

import { Observable, combineLatest, BehaviorSubject, of } from 'rxjs'

// insertion point sub template for services imports{{` + string(rune(NgLibFrontRepoServiceImports)) + `}}

import { BackRepoData } from './back-repo-data'

export const StackType = "{{PkgPathRoot}}/models"

// FrontRepo stores all instances in a front repository (design pattern repository)
export class FrontRepo { // insertion point sub template{{` + string(rune(NgLibFrontRepoMapDecl)) + `}}

	// getFrontArray allows for a get function that is robust to refactoring of the named struct name
	// for instance frontRepo.getArray<Astruct>( Astruct.GONGSTRUCT_NAME), is robust to a refactoring of Astruct identifier
	// contrary to frontRepo.Astructs_array which is not refactored when Astruct identifier is modified
	getFrontArray<Type>(gongStructName: string): Array<Type> {
		switch (gongStructName) {
			// insertion point{{` + string(rune(NgLibFrontRepoSwitchGetFrontArray)) + `}}
			default:
				throw new Error("Type not recognized");
		}
	}

	getFrontMap<Type>(gongStructName: string): Map<number, Type> {
		switch (gongStructName) {
			// insertion point{{` + string(rune(NgLibFrontRepoSwitchGetFrontMap)) + `}}
			default:
				throw new Error("Type not recognized");
		}
	}
}

// the table component is called in different ways
//
// DISPLAY or ASSOCIATION MODE
//
// in ASSOCIATION MODE, it is invoked within a diaglo and a Dialog Data item is used to
// configure the component
// DialogData define the interface for information that is forwarded from the calling instance to 
// the select table
export class DialogData {
	ID: number = 0 // ID of the calling instance

	// the reverse pointer is the name of the generated field on the destination
	// struct of the ONE-MANY association
	ReversePointer: string = "" // field of {{Structname}} that serve as reverse pointer
	OrderingMode: boolean = false // if true, this is for ordering items

	// there are different selection mode : ONE_MANY or MANY_MANY
	SelectionMode: SelectionMode = SelectionMode.ONE_MANY_ASSOCIATION_MODE

	// used if SelectionMode is MANY_MANY_ASSOCIATION_MODE
	//
	// In Gong, a MANY-MANY association is implemented as a ONE-ZERO/ONE followed by a ONE_MANY association
	// 
	// in the MANY_MANY_ASSOCIATION_MODE case, we need also the Struct and the FieldName that are
	// at the end of the ONE-MANY association
	SourceStruct: string = ""	// The "Aclass"
	SourceField: string = "" // the "AnarrayofbUse"
	IntermediateStruct: string = "" // the "AclassBclassUse" 
	IntermediateStructField: string = "" // the "Bclass" as field
	NextAssociationStruct: string = "" // the "Bclass"

	GONG__StackPath: string = ""
}

export enum SelectionMode {
	ONE_MANY_ASSOCIATION_MODE = "ONE_MANY_ASSOCIATION_MODE",
	MANY_MANY_ASSOCIATION_MODE = "MANY_MANY_ASSOCIATION_MODE",
}

//
// observable that fetch all elements of the stack and store them in the FrontRepo
//
@Injectable({
	providedIn: 'root'
})
export class FrontRepoService {

	GONG__StackPath: string = ""
	private socket: WebSocket | undefined

	httpOptions = {
		headers: new HttpHeaders({ 'Content-Type': 'application/json' })
	};

	//
	// Store of all instances of the stack
	//
	frontRepo = new (FrontRepo)

	constructor(
		private http: HttpClient, // insertion point sub template {{` + string(rune(NgLibFrontRepoServiceDecl)) + `}}
	) { }

	// postService provides a post function for each struct name
	postService(structName: string, instanceToBePosted: any) {
		let service = this[structName.toLowerCase() + "Service" + "Service" as keyof FrontRepoService]
		let servicePostFunction = service[("post" + structName) as keyof typeof service] as (instance: typeof instanceToBePosted) => Observable<typeof instanceToBePosted>

		servicePostFunction(instanceToBePosted).subscribe(
			instance => {
				let behaviorSubject = instanceToBePosted[(structName + "ServiceChanged") as keyof typeof instanceToBePosted] as unknown as BehaviorSubject<string>
				behaviorSubject.next("post")
			}
		);
	}

	// deleteService provides a delete function for each struct name
	deleteService(structName: string, instanceToBeDeleted: any) {
		let service = this[structName.toLowerCase() + "Service" as keyof FrontRepoService]
		let serviceDeleteFunction = service["delete" + structName as keyof typeof service] as (instance: typeof instanceToBeDeleted) => Observable<typeof instanceToBeDeleted>

		serviceDeleteFunction(instanceToBeDeleted).subscribe(
			instance => {
				let behaviorSubject = instanceToBeDeleted[(structName + "ServiceChanged") as keyof typeof instanceToBeDeleted] as unknown as BehaviorSubject<string>
				behaviorSubject.next("delete")
			}
		);
	}

	// typing of observable can be messy in typescript. Therefore, one force the type
	observableFrontRepo: [
		Observable<null>, // see below for the of(null) observable
		// insertion point sub template {{` + string(rune(NgLibFrontRepoObservableArrayType)) + `}}
	] = [
			// Using "combineLatest" with a placeholder observable.
			//
			// This allows the typescript compiler to pass when no GongStruct is present in the front API
			//
			// The "of(null)" is a "meaningless" observable that emits a single value (null) and completes.
			// This is used as a workaround to satisfy TypeScript requirements and the "combineLatest" 
			// expectation for a non-empty array of observables.
			of(null), // 
			// insertion point sub template{{` + string(rune(NgLibFrontRepoObservableRefs)) + `}}
		];

	//
	// pull performs a GET on all struct of the stack and redeem association pointers 
	//
	// This is an observable. Therefore, the control flow forks with
	// - pull() return immediatly the observable
	// - the observable observer, if it subscribe, is called when all GET calls are performs
	pull(GONG__StackPath: string = ""): Observable<FrontRepo> {

		this.GONG__StackPath = GONG__StackPath

		this.observableFrontRepo = [
			of(null), // see above for justification
			// insertion point sub template{{` + string(rune(NgLibFrontRepoObservableRefs)) + `}}
		]

		return new Observable<FrontRepo>(
			(observer) => {
				combineLatest(
					this.observableFrontRepo
				).subscribe(
					([
						___of_null, // see above for the explanation about of
						// insertion point sub template for declarations {{` + string(rune(NgLibFrontRepoArraysDecls)) + `}}
					]) => {
						let _this = this
						// Typing can be messy with many items. Therefore, type casting is necessary here
						// insertion point sub template for type casting {{` + string(rune(NgLibFrontRepoTypeCasting)) + `}}

						// 
						// First Step: init map of instances
						// insertion point sub template for init {{` + string(rune(NgLibFrontRepoInitMapInstances)) + `}}

						// 
						// Second Step: reddeem front objects
						// insertion point sub template for redeem {{` + string(rune(NgLibFrontRepoInitFrontObjects)) + `}}

						// hand over control flow to observer
						observer.next(this.frontRepo)
					}
				)
			}
		)
	}

	public connectToWebSocket(GONG__StackPath: string): Observable<FrontRepo> {

		this.GONG__StackPath = GONG__StackPath


		let params = new HttpParams().set("GONG__StackPath", this.GONG__StackPath)
		let basePath = 'ws://localhost:8080/api/{{PkgPathRoot}}/v1/ws/stage'
		let paramString = params.toString()
		let url = ` + "`" + "${basePath}?${paramString}" + "`" + `
		this.socket = new WebSocket(url)

		return new Observable(observer => {
			this.socket!.onmessage = event => {


				const backRepoData = new BackRepoData(JSON.parse(event.data))

				let frontRepo = new (FrontRepo)

				// 
				// First Step: init map of instances
				// insertion point sub template for init 
				// init the arrays
				// insertion point sub template for init {{` + string(rune(NgLibFrontRepoInitMapInstancesFromWebSocket)) + `}}

				// 
				// Second Step: reddeem front objects
				// insertion point sub template for redeem 
				// fill up front objects
				// insertion point sub template for redeem {{` + string(rune(NgLibFrontRepoInitFrontObjectsFromWebSocket)) + `}}


				observer.next(this.frontRepo)
			}
			this.socket!.onerror = event => {
				observer.error(event)
			}
			this.socket!.onclose = event => {
				observer.complete()
			}

			return () => {
				this.socket!.close()
			}
		})
	}
}

// insertion point for get unique ID per struct {{` + string(rune(NgLibFrontRepoPerStructGetUniqueID)) + `}}
`

type FrontRepoInsertionPointId int

const (
	NgLibFrontRepoServiceImports FrontRepoInsertionPointId = iota
	NgLibFrontRepoMapDecl
	NgLibFrontRepoSwitchGetFrontArray
	NgLibFrontRepoSwitchGetFrontMap
	NgLibFrontRepoObservableArrayType
	NgLibFrontRepoTypeCasting
	NgLibFrontRepoServiceDecl
	NgLibFrontRepoObservableRefs
	NgLibFrontRepoArraysDecls
	NgLibFrontRepoInitMapInstances
	NgLibFrontRepoInitMapInstancesFromWebSocket
	NgLibFrontRepoRedeemPointers
	NgLibFrontRepoInitFrontObjects
	NgLibFrontRepoInitFrontObjectsFromWebSocket
	NgLibFrontRepoSlicesOfPointersDecode
	NgLibFrontRepoPerStructGetUniqueID
)

var NgFrontRepoPerStructTmplCodes map[FrontRepoInsertionPointId]string = // new line
map[FrontRepoInsertionPointId]string{

	NgLibFrontRepoServiceImports: `
import { {{Structname}}API } from './{{structname}}-api'
import { {{Structname}}, Copy{{Structname}}APITo{{Structname}} } from './{{structname}}'
import { {{Structname}}Service } from './{{structname}}.service'
`,

	NgLibFrontRepoMapDecl: `
	array_{{Structname}}s = new Array<{{Structname}}>() // array of front instances
	map_ID_{{Structname}} = new Map<number, {{Structname}}>() // map of front instances
`,

	NgLibFrontRepoSwitchGetFrontArray: `
			case '{{Structname}}':
				return this.array_{{Structname}}s as unknown as Array<Type>`,

	NgLibFrontRepoSwitchGetFrontMap: `
			case '{{Structname}}':
				return this.map_ID_{{Structname}} as unknown as Map<number, Type>`,

	NgLibFrontRepoObservableArrayType: `
		Observable<{{Structname}}API[]>,`,

	NgLibFrontRepoServiceDecl: `
		private {{structname}}Service: {{Structname}}Service,`,

	NgLibFrontRepoObservableRefs: `
			this.{{structname}}Service.get{{Structname}}s(this.GONG__StackPath, this.frontRepo),`,

	NgLibFrontRepoArraysDecls: `
						{{structname}}s_,`,

	NgLibFrontRepoTypeCasting: `
						var {{structname}}s: {{Structname}}API[]
						{{structname}}s = {{structname}}s_ as {{Structname}}API[]`,

	NgLibFrontRepoInitMapInstances: `
						// init the arrays
						this.frontRepo.array_{{Structname}}s = []
						this.frontRepo.map_ID_{{Structname}}.clear()

						{{structname}}s.forEach(
							{{structname}}API => {
								let {{structname}} = new {{Structname}}
								this.frontRepo.array_{{Structname}}s.push({{structname}})
								this.frontRepo.map_ID_{{Structname}}.set({{structname}}API.ID, {{structname}})
							}
						)
`,

	NgLibFrontRepoInitMapInstancesFromWebSocket: `
				// init the arrays
				frontRepo.array_{{Structname}}s = []
				frontRepo.map_ID_{{Structname}}.clear()

				backRepoData.{{Structname}}APIs.forEach(
					{{structname}}API => {
						let {{structname}} = new {{Structname}}
						frontRepo.array_{{Structname}}s.push({{structname}})
						frontRepo.map_ID_{{Structname}}.set({{structname}}API.ID, {{structname}})
					}
				)
`,

	NgLibFrontRepoInitFrontObjects: `
						// fill up front objects
						{{structname}}s.forEach(
							{{structname}}API => {
								let {{structname}} = this.frontRepo.map_ID_{{Structname}}.get({{structname}}API.ID)
								Copy{{Structname}}APITo{{Structname}}({{structname}}API, {{structname}}!, this.frontRepo)
							}
						)
`,

	NgLibFrontRepoInitFrontObjectsFromWebSocket: `
				// fill up front objects
				backRepoData.{{Structname}}APIs.forEach(
					{{structname}}API => {
						let {{structname}} = frontRepo.map_ID_{{Structname}}.get({{structname}}API.ID)
						Copy{{Structname}}APITo{{Structname}}({{structname}}API, {{structname}}!, frontRepo)
					}
				)
`,

	NgLibFrontRepoRedeemPointers: `
						{{structname}}s.forEach(
							{{structname}} => {
								// insertion point sub sub template for ONE-/ZERO-ONE associations pointers redeeming{{` + string(rune(NgFrontRepoPtrToStructRedeeming)) + `}}
							}
						)`,

	NgLibFrontRepoSlicesOfPointersDecode: `
						{{structname}}s.forEach(
							{{structname}} => {
								// insertion point sub sub template for ONE-/ZERO-ONE associations pointers redeeming{{` + string(rune(NgFrontRepoPtrToStructRedeeming)) + `}}
								// insertion point for pointers decoding{{` + string(rune(NgFrontRepoSliceOfPointerSorting)) + `}}
							}
						)`,

	NgLibFrontRepoPerStructGetUniqueID: `
export function get{{Structname}}UniqueID(id: number): number {
	return {{Prime}} * id
}`,
}

type NgLibFrontRepoServiceSubSubTemplate int

const (
	NgFrontRepoPtrToStructRedeeming NgLibFrontRepoServiceSubSubTemplate = iota + 1789
	NgFrontRepoSliceOfPointerSorting
)

// for each sub sub template, what sub template it relates to
var NgLibFrontRepoPointerToStructSubSubToSubMap map[NgLibFrontRepoServiceSubSubTemplate]string = //
map[NgLibFrontRepoServiceSubSubTemplate]string{
	NgFrontRepoPtrToStructRedeeming: string(rune(NgLibFrontRepoRedeemPointers)),
}

var NgFrontRepoPtrToStructTmplCodes map[NgLibFrontRepoServiceSubSubTemplate]string = // new line
map[NgLibFrontRepoServiceSubSubTemplate]string{
	NgFrontRepoPtrToStructRedeeming: `
								// insertion point for pointer field {{FieldName}} redeeming
								{
									let _{{assocStructName}} = this.frontRepo.{{AssocStructName}}s.get({{structname}}.{{Structname}}PointersEncoding.{{FieldName}}ID.Int64)
									if (_{{assocStructName}}) {
										{{structname}}.{{FieldName}} = _{{assocStructName}}
									}
								}`,
	NgFrontRepoSliceOfPointerSorting: `
								{{structname}}.{{FieldName}} = new Array<{{AssocStructName}}API>()
								for (let _id of {{structname}}.{{Structname}}PointersEncoding.{{FieldName}}) {
									let _{{assocStructName}} = this.frontRepo.{{AssocStructName}}s.get(_id)
									if (_{{assocStructName}} != undefined) {
										{{structname}}.{{FieldName}}.push(_{{assocStructName}}!)
									}
								}`,
}

func CodeGeneratorNgFrontRepo(modelPkg *models.ModelPkg) {

	pkgName := modelPkg.Name
	matTargetPath := modelPkg.NgDataLibrarySourceCodeDirectory
	pkgGoPath := modelPkg.PkgPath

	file, err := os.Create(filepath.Join(matTargetPath, "front-repo.service.ts"))
	if err != nil {
		log.Panic(err)
	}

	code := NgLibFrontRepoServiceTemplate

	perStructCodes := make(map[FrontRepoInsertionPointId]string)
	for subTemplate := range NgFrontRepoPerStructTmplCodes {
		perStructCodes[subTemplate] = ""
	}

	// have alphabetical order generation
	structList := []*models.GongStruct{}
	for _, _struct := range modelPkg.GongStructs {
		if !_struct.HasNameField() || _struct.IsIgnoredForFront {
			continue
		}
		structList = append(structList, _struct)
	}
	sort.Slice(structList[:], func(i, j int) bool {
		return structList[i].Name < structList[j].Name
	})

	for structIndex, _struct := range structList {

		structName := strings.ToLower(_struct.Name)
		_ = structName

		fieldPointerToStructCodes := ""
		fieldSliceOfPtrSortingToStructCodes := ""

		// compute code per field
		for _, field := range _struct.Fields {
			switch field := field.(type) {
			case *models.PointerToGongStructField:

				fieldName := strings.ToLower(field.Name)
				assocStructName := strings.ToLower(field.GongStruct.Name)

				fieldPointerToStructCodes += models.Replace6(
					NgFrontRepoPtrToStructTmplCodes[NgFrontRepoPtrToStructRedeeming],
					"{{fieldName}}", fieldName,
					"{{FieldName}}", field.Name,
					"{{AssocStructName}}", field.GongStruct.Name,
					"{{assocStructName}}", assocStructName,
					"{{Structname}}", _struct.Name,
					"{{structname}}", structName)

			case *models.SliceOfPointerToGongStructField:

				fieldName := strings.ToLower(field.Name)
				assocStructName := strings.ToLower(field.GongStruct.Name)

				fieldSliceOfPtrSortingToStructCodes += models.Replace6(
					NgFrontRepoPtrToStructTmplCodes[NgFrontRepoSliceOfPointerSorting],
					"{{fieldName}}", fieldName,
					"{{FieldName}}", field.Name,
					"{{AssocStructName}}", field.GongStruct.Name,
					"{{assocStructName}}", assocStructName,
					"{{Structname}}", _struct.Name,
					"{{structname}}", structName)
			}
		}

		// compute code from sub template
		for subTemplate := range NgFrontRepoPerStructTmplCodes {
			perStructCodes[subTemplate] += models.Replace3(NgFrontRepoPerStructTmplCodes[subTemplate],
				"{{Structname}}", _struct.Name,
				"{{structname}}", structName,
				"{{Prime}}", strconv.Itoa(models.Primes[structIndex+10]))

			toReplace := "{{" + string(rune(NgFrontRepoPtrToStructRedeeming)) + "}}"
			perStructCodes[subTemplate] = strings.ReplaceAll(
				perStructCodes[subTemplate],
				toReplace, fieldPointerToStructCodes)

			toReplace = "{{" + string(rune(NgFrontRepoSliceOfPointerSorting)) + "}}"
			perStructCodes[subTemplate] = strings.ReplaceAll(
				perStructCodes[subTemplate],
				toReplace, fieldSliceOfPtrSortingToStructCodes)
		}
	}

	//
	// final per struct replacement
	//
	for perStructTmplCode := range NgFrontRepoPerStructTmplCodes {
		toReplace := "{{" + string(rune(perStructTmplCode)) + "}}"
		code = strings.ReplaceAll(code, toReplace, perStructCodes[perStructTmplCode])
	}

	code = models.Replace4(code,
		"{{PkgName}}", pkgName,
		"{{TitlePkgName}}", strings.Title(pkgName),
		"{{pkgname}}", strings.ToLower(pkgName),
		"{{PkgPathRoot}}", strings.ReplaceAll(pkgGoPath, "/models", ""))

	defer file.Close()
	fmt.Fprint(file, code)
}
